### PR TITLE
New spell: Projected Weapon (L5 Translocations)

### PR DESCRIPTION
**Pull request by PleasingFungus**
Projected Weapon launches attacks with your melee weapon against 2-4
random enemies in LOS, depending on spellpower. It's themed as opening
one "in" portal and a bunch of tiny "out" portals, letting you hit many
enemies with a single attack.

(More exciting name wanted!)

This spell is intended to give melee-focused characters more reason to
consider investing in spellcasting, It's fun to try to skill toward a
'hybrid' build, and a spell that mostly scales off your melee attack,
rather than your spellpower, helps encourage that. This being a L5
spell helps avoid it being a truly free 'gimme', and allows it to hit
many enemies at once.

The lack of targeter should hopefully avoid it being annoying to use,
by making it just a single press to cast. It also gives the player
more reason to keep normally hitting things in melee sometimes,
when it's important to kill a specific target.

UC attacks are banned to avoid this being too cheap for transmuters,
which already have light armour and spellcasting skills. That could be
the wrong call!

Cleaving also doesn't work with this spell, since it seems overly
confusing, and since reach and riposte don't work with it either.
(Thus, it's 'fair'.)

### CODE REVIEW

**Review by ebering — CHANGES_REQUESTED** [review]
This spell is a neat idea, but I don't think it should be more expensive than passage of Golubria, probably. The air dual school I'm skeptical of for reasons elaborated on in the comment.

There's a lot more technical work that needs to be done to get this thing online, here's my assessment of quirks, thorns, and bugs.

### Sleep stabs & things

Granting stabs with this spell does seem too easy; I think distraction stabs should also not be granted, since reach, riposte, and cleave are also out. Beyond that, the spell should be a normal weapon swing:

- Should use normal attack delay (currently tries to but fails, see first bug)
- Should trigger all weapon effects (except cleaving) (currently does, except for G&G)
- Should fumble in water, be affected by halos and umbras and coronas, etc. (all of this works correctly).

### Attack delay depends on the number of targets

Augh. `melee_attack::handle_phase_attempted` sets `you.time_taken` to `you.attack_delay()` (melee-attack.cc L209; however `you.attack_delay()` uses the current value of `you.time_taken` (see player-act.cc L297 and the preceding TODO comment there).

A solution would be to have melee_attack not set the time for projected attacks and roll the attack delay in projected weapon. If you have the energy diving that TODO is also a solution, and a third solution that might fix other quirks is below.

### Bugs involving the interaction with cleaving:

- Axes will cleave to monsters *adjacent to the caster* for each projected attack.
- Gyre & Gimble don't get their second attack (except they do against an adjacent monster)

### "Bugs" involving the commit message not matching the code:

- The commit message says sleep stabs are forbidden but there's nothing in the code and they work normally.

### The interaction with unrandarts isn't great

The following unrands do things that might cause penance even if the attack target is ok: Sword of Power (chance to fire a sword beam at *each* target), Devastator (explodes *each* target), Mace of Variability (chance for a chaos arc at *each* target), Singing Sword (sonic wave roll for *each* target), Sceptre of Torment (torment chance for *each* target), Arc Blade discharges. Thankfully, this doesn't lead to unprompted penance.

What happens is the player is given a stop_attack_prompt for each, allowing selective application of the not-full-los weaponspells, as well as a pain interface-wise. Better might look like:

> Go through the checks in melee-attack.cc lines 105-185 (re-factoring them to their own place, this was a suggested project part or starting place for @neilmoore 's students, not sure if they've taken it up but if so it looks like we've generated a requirement for them to gather!) for each potential target, abort the whole spell on cancel and don't repeat the prompts if the player confirms they really want to cast the spell.

That's got its own issues: 
- interface-wise: it makes it hard to cast the spell in a highly-likely penance situation if you really mean it (maybe ok?)
- technically it looks like it'll be a big pain because it requires breaking melee_attack's encapsulation, oof.

A different possible approach:

> Set up the targets as "cleave" targets, so that a single melee_attack handles all the prompting and cancellations, that way if the attack cancels you can cancel the spell.

Also has its own issues:
- Getting fail_check() into the place between when the prompts happen and when the attack fires seems hard with the current code structure.
- Upon further digging it looks like the attack prompts for cleaving targets aren't checked for unrands with funny effects. So this is probably not the way to go. (Getting Cleave from Xom wrath while using devastator or the sword of power seems like a way to get unprompted penance, what an edge case.)

**Comment by ebering** [comment]
Separate name bike shed/brainstorm comment:

- Warp Weapon
- Phase Shift
- Displaced Thrust
- Spatial Echo
- Wandering Blade
- Discursivity
- Blessing of the Void

File: crawl-ref/source/book-data.h [hunk 1]
```diff
@@ -241,6 +243,7 @@ static const vector<spell_type> spellbook_templates[] =
 #endif
 
 {   // Book of Power
+    SPELL_LRD,
```
[the remark this answers is not in the document]
Comment (PleasingFungus) [reply crawl-ref/source/book-data.h]: Whoops! Leftover from an earlier revision where i removed LRD from book of burglary.

File: crawl-ref/source/book-data.h [hunk 2]
```diff
@@ -71,6 +71,7 @@ static const vector<spell_type> spellbook_templates[] =
     SPELL_GRAVITAS,
     SPELL_TELEPORT_OTHER,
     SPELL_GOLUBRIAS_PASSAGE,
```
[the remark this answers is not in the document]
Comment (PleasingFungus) [reply crawl-ref/source/book-data.h]: Sure, I'll move PoG to Book of the Warp.

File: crawl-ref/source/spl-data.h [hunk 3]
```diff
@@ -3363,6 +3363,17 @@ static const struct spell_desc spelldata[] =
     TILEG_GENERIC_MONSTER_SPELL
 },
 
+{
+    SPELL_PROJECTED_WEAPON, "Transdimensional Blitz",
+    spschool::translocation | spschool::air,
```
[the remark this answers is not in the document]
Comment (PleasingFungus) [reply crawl-ref/source/spl-data.h]: Sure, I'm fine switching it back. Air school was a request from Hellmonk on the principle that pure tloc shouldn't do damage, but gammafunk argued that pproj was a reasonable analogy here.

**Review by PleasingFungus — COMMENTED** [review]
Thanks for the in-depth review! I'll take a pass at your top-level comments shortly.

**Comment by PleasingFungus** [comment]
> The commit message says sleep stabs are forbidden but there's nothing in the code and they work normally.

For what it's worth, I was unable to replicate this bug. See this section:
```
        // manually apply noise
        behaviour_event(targets[i], ME_ALERT, &you, you.pos()); // shout + set you as foe

        melee_attack atk(&you, targets[i]);
        atk.is_projected = true;
        atk.attack();
```

The `ME_ALERT` is the same logic that prevents palentongas from roll-stabbing sleeping enemies - it wakes them up first.

I'm personally undecided about whether we want to prevent all stabs, since it seems like a neat and reasonably fair combo to petrify -> this, etc. For now I've disabled it, but I'm open to discussion.

**Comment by PleasingFungus** [comment]
OK, think I've addressed everything. Thanks again for doing all this testing!

I'm leaning toward re-enabling non-sleep stabs, since it seems fine to allow a level 5 spell + a hex effect to allow stabbing at range, but am open to discussion.

File: crawl-ref/source/spl-transloc.cc [hunk 4]
```diff
@@ -1044,6 +1047,90 @@ spret cast_portal_projectile(int pow, bool fail)
...
+
+    shuffle_array(targets);
+    const size_t max_targets = 2 + div_rand_round(pow, 50);
+    for (size_t i = 0; i < max_targets && i < targets.size(); i++)
+    {
+        // manually apply noise
+        behaviour_event(targets[i], ME_ALERT, &you, you.pos()); // shout + set you as foe
+
+        // Somewhat hacky: reset attack delay before each attack so that only the final
+        // attack ends up actually setting time taken. (No quadratic effects.)
```
Comment (ebering) [inline crawl-ref/source/spl-transloc.cc:1120]: Maybe include a mention of the XXX in attack_delay, so if someone fixes that there's hope that this hack gets removed too

File: crawl-ref/source/spl-transloc.cc [hunk 5]
```diff
@@ -1044,6 +1047,90 @@ spret cast_portal_projectile(int pow, bool fail)
...
+    mpr("Space warps momentarily into an impossible shape!");
+
+    const int initial_time = you.time_taken;
+
+    shuffle_array(targets);
+    const size_t max_targets = 2 + div_rand_round(pow, 50);
+    for (size_t i = 0; i < max_targets && i < targets.size(); i++)
+    {
+        // manually apply noise
+        behaviour_event(targets[i], ME_ALERT, &you, you.pos()); // shout + set you as foe
```
Comment (ebering) [inline crawl-ref/source/spl-transloc.cc:1117]: Is this necessary when melee_attack disables stabs?

**Review by ebering — APPROVED** [review]
Looks good now, two small comments. 

I'm still in the "no stabs" camp with this. The fun of stabbing is the thrill of the approach, or of the risk when the target wakes up and now must be hexed and approached; this would cheapen that thrill. This is a separate additional reason that's more compelling to me than the "consistency with other weapon specials".

**Comment by PleasingFungus** [comment]
> I'm still in the "no stabs" camp with this. The fun of stabbing is the thrill of the approach, or of the risk when the target wakes up and now must be hexed and approached; this would cheapen that thrill. This is a separate additional reason that's more compelling to me than the "consistency with other weapon specials".

Sure, the thrill of the approach is definitely big fun - I see your argument. I was leaning toward "it's fun to give stabbers new tools and strategic choices", but maybe it'd be a bit too cheap and easy. I'm fine with going stab-free for now.

File: crawl-ref/source/spl-transloc.cc (hunk 4 above)
Comment (PleasingFungus) [reply crawl-ref/source/spl-transloc.cc:1120]: Good idea. Done.

**Pull request merged**